# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/jit-test/tests/gc/finalizationGroup.js
# Commit: 8e609f97c413
# Full Hash: 8e609f97c413a12f7042dfe80647bdd34f0400d8
# Author: Jon Coppeard <jcoppeard@mozilla.com>
# Date: 2019-11-01 16:13:11
# Regressor Bug: 1587096
# File Overlap Count: 1
# Description:
#   Bug 1587096 - Part 1: Add JS FinalizationGroupObject and related data structures r=sfink
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D49943
# ==============================================================================

diff -r f273ec2ec0ae -r 8e609f97c413 js/src/jit-test/tests/gc/finalizationGroup.js
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/js/src/jit-test/tests/gc/finalizationGroup.js	Fri Nov 01 10:37:25 2019 +0000
@@ -0,0 +1,44 @@
+// |jit-test| --enable-weak-refs
+
+function checkPropertyDescriptor(obj, property, writable, enumerable,
+                                 configurable) {
+    let desc = Object.getOwnPropertyDescriptor(obj, property);
+    assertEq(typeof desc, "object");
+    assertEq(desc.writable, writable);
+    assertEq(desc.enumerable, enumerable);
+    assertEq(desc.configurable, configurable);
+}
+
+function assertThrowsTypeError(thunk) {
+    let error;
+    try {
+        thunk();
+    } catch (e) {
+        error = e;
+    }
+    assertEq(error instanceof TypeError, true);
+}
+
+// 3.1 The FinalizationGroup Constructor
+assertEq(typeof this.FinalizationGroup, "function");
+
+// 3.2 Properties of the FinalizationGroup Constructor
+assertEq(Object.getPrototypeOf(FinalizationGroup), Function.prototype);
+
+// 3.2.1 FinalizationGroup.prototype
+checkPropertyDescriptor(FinalizationGroup, 'prototype', false, false, false);
+
+// 3.3 Properties of the FinalizationGroup Prototype Object
+let proto = FinalizationGroup.prototype;
+assertEq(Object.getPrototypeOf(proto), Object.prototype);
+
+// 3.3.1 FinalizationGroup.prototype.constructor
+assertEq(proto.constructor, FinalizationGroup);
+
+//3.3.5 FinalizationGroup.prototype [ @@toStringTag ]
+assertEq(proto[Symbol.toStringTag], "FinalizationGroup");
+checkPropertyDescriptor(proto, Symbol.toStringTag, false, false, true);
+
+assertThrowsTypeError(() => new FinalizationGroup());
+assertThrowsTypeError(() => new FinalizationGroup(1));
+let group = new FinalizationGroup(x => 1);